File: backend/internal/labels/studio.go
package labels

// The templates, and the data a print run needs (blueprint B3).

import (
	"context"
	"encoding/json"
	"errors"
	"fmt"
	"strings"
	"time"

	"github.com/google/uuid"
	"github.com/jackc/pgx/v5"
	"github.com/shopspring/decimal"

	"github.com/mahedi-emon/rawsyst-pos/backend/internal/platform/audit"
	"github.com/mahedi-emon/rawsyst-pos/backend/internal/platform/db"
	"github.com/mahedi-emon/rawsyst-pos/backend/internal/platform/errs"
)

// Templates lists the layouts, defaults first within each kind.
func (s *Service) Templates(ctx context.Context, scope Scope) ([]Template, error) {
	out := []Template{}
	err := s.pool.TxAsTenant(ctx, scope.TenantID, func(tx pgx.Tx) error {
		rows, e := tx.Query(ctx, templateSelect+`
			WHERE t.company_id = $1
			ORDER BY t.kind, t.is_default DESC, t.name`, scope.CompanyID)
		if e != nil {
			return e
		}
		defer rows.Close()
		for rows.Next() {
			tpl, e := scanTemplate(rows)
			if e != nil {
				return e
			}
			out = append(out, tpl)
		}
		return rows.Err()
	})
	return out, db.Translate(err, "")
}

// SaveTemplate creates or replaces a layout.
func (s *Service) SaveTemplate(
	ctx context.Context, scope Scope, id *uuid.UUID, in Template,
) (Template, error) {
	in.Name = strings.TrimSpace(in.Name)
	if in.Name == "" {
		return Template{}, errs.Validation("Give the label a name.").
			WithField("name", "It is what the print button will offer.")
	}
	if !knownKind(in.Kind) {
		return Template{}, errs.Newf(errs.CodeInvalidInput,
			"There is no kind of label called %q.", in.Kind)
	}
	width, err := decimal.NewFromString(strings.TrimSpace(in.Width))
	if err != nil || !width.IsPositive() {
		return Template{}, errs.Validation("Say how wide the label is.").
			WithField("width_mm", "In millimetres, as printed on the roll.")
	}
	height, err := decimal.NewFromString(strings.TrimSpace(in.Height))
	if err != nil || !height.IsPositive() {
		return Template{}, errs.Validation("Say how tall the label is.").
			WithField("height_mm", "In millimetres, as printed on the roll.")
	}
	if in.Kind == KindA4Sheet && (in.Columns == nil || in.Rows == nil) {
		return Template{}, errs.Validation(
			"An A4 sheet needs to say how the labels sit on it.").
			WithField("columns", "Three across and eight down is a sheet of 24.")
	}
	if in.Kind != KindA4Sheet {
		// A single label carrying a grid would leave the renderer choosing
		// between two contradictory layouts. The database refuses it too.
		in.Columns, in.Rows = nil, nil
	}
	if err := checkFields(in.Fields); err != nil {
		return Template{}, err
	}
	if len(in.Fields) == 0 {
		return Template{}, errs.Validation(
			"A label with nothing on it is a blank sticker.").
			WithField("fields", "Choose at least one thing to print.")
	}

	fields, err := json.Marshal(in.Fields)
	if err != nil {
		return Template{}, err
	}
	margin := amountOr(in.Margin)
	gap := amountOr(in.Gap)

	var saved uuid.UUID
	err = s.pool.TxAsTenant(ctx, scope.TenantID, func(tx pgx.Tx) error {
		// The default is cleared first. The unique index would otherwise refuse
		// the write, and "this label cannot be saved" is a worse answer than
		// moving a default the person plainly intended to move.
		if in.IsDefault {
			if _, e := tx.Exec(ctx, `
				UPDATE label_template SET is_default = false
				WHERE company_id = $1 AND kind = $2 AND is_default`,
				scope.CompanyID, in.Kind); e != nil {
				return e
			}
		}

		if id != nil {
			tag, e := tx.Exec(ctx, `
				UPDATE label_template
				SET name = $3, kind = $4, width_mm = $5, height_mm = $6,
				    columns = $7, rows = $8, margin_mm = $9, gap_mm = $10,
				    fields = $11::jsonb, is_default = $12
				WHERE id = $1 AND company_id = $2`,
				*id, scope.CompanyID, in.Name, in.Kind, width, height,
				in.Columns, in.Rows, margin, gap, string(fields), in.IsDefault)
			if e != nil {
				return db.Translate(e, "That label could not be saved.")
			}
			if tag.RowsAffected() == 0 {
				return errs.New(errs.CodeNotFound, "That label was not found.")
			}
			saved = *id
			return auditTemplate(ctx, tx, scope, "label_template_changed", saved,
				nil, describeTemplate(in, string(fields)))
		}

		if e := db.Translate(tx.QueryRow(ctx, `
			INSERT INTO label_template
			  (tenant_id, company_id, name, kind, width_mm, height_mm, columns,
			   rows, margin_mm, gap_mm, fields, is_default, created_by)
			VALUES ($1,$2,$3,$4,$5,$6,$7,$8,$9,$10,$11::jsonb,$12,$13)
			RETURNING id`,
			scope.TenantID, scope.CompanyID, in.Name, in.Kind, width, height,
			in.Columns, in.Rows, margin, gap, string(fields), in.IsDefault,
			scope.UserID).Scan(&saved),
			"A label with that name already exists."); e != nil {
			return e
		}
		return auditTemplate(ctx, tx, scope, "label_template_created", saved,
			nil, describeTemplate(in, string(fields)))
	})
	if err != nil {
		return Template{}, db.Translate(err, "")
	}
	return s.Template(ctx, scope, saved)
}

// auditTemplate records a change to a layout.
//
// The tag carries the shop's price, and changing what is on it changes every
// ticket printed from then on. The bulk barcode generator beside it has been
// audited since 0095 and this was not, so a template whose price line had been
// taken off named nobody.
func auditTemplate(
	ctx context.Context, tx pgx.Tx, scope Scope, action string,
	id uuid.UUID, before, after map[string]any,
) error {
	return audit.Write(ctx, tx, audit.Entry{
		TenantID: &scope.TenantID, ActorID: &scope.UserID,
		ActorLabel: audit.LabelFor(ctx, tx, scope.UserID),
		Action:     action,
		EntityType: "label_template", EntityID: &id,
		Before: before, After: after,
	})
}

// Template reads one.
func (s *Service) Template(
	ctx context.Context, scope Scope, id uuid.UUID,
) (Template, error) {
	var out Template
	err := s.pool.TxAsTenant(ctx, scope.TenantID, func(tx pgx.Tx) error {
		row := tx.QueryRow(ctx, templateSelect+`
			WHERE t.id = $1 AND t.company_id = $2`, id, scope.CompanyID)
		tpl, e := scanTemplate(row)
		if e == pgx.ErrNoRows {
			return errs.New(errs.CodeNotFound, "That label was not found.")
		}
		out = tpl
		return e
	})
	return out, db.Translate(err, "")
}

// DeleteTemplate removes a layout.
func (s *Service) DeleteTemplate(
	ctx context.Context, scope Scope, id uuid.UUID,
) error {
	return db.Translate(s.pool.TxAsTenant(ctx, scope.TenantID,
		func(tx pgx.Tx) error {
			// Read before removing, so the trail says WHICH layout went. A row
			// naming a uuid that no longer resolves tells an investigation
			// nothing at all.
			var name, kind string
			var wasDefault bool
			e := tx.QueryRow(ctx, `
				SELECT name, kind, is_default FROM label_template
				WHERE id = $1 AND company_id = $2`, id, scope.CompanyID).
				Scan(&name, &kind, &wasDefault)
			if errors.Is(e, pgx.ErrNoRows) {
				return errs.New(errs.CodeNotFound, "That label was not found.")
			}
			if e != nil {
				return e
			}

			if _, e := tx.Exec(ctx, `
				DELETE FROM label_template WHERE id = $1 AND company_id = $2`,
				id, scope.CompanyID); e != nil {
				return db.Translate(e,
					"That label cannot be removed while something still uses it.")
			}
			return auditTemplate(ctx, tx, scope, "label_template_removed", id,
				map[string]any{
					"name": name, "kind": kind, "was_default": wasDefault,
				}, nil)
		}), "")
}

// describeTemplate is what an audit entry records about a layout.
//
// The field list is written out rather than counted: "6 fields" says nothing,
// and the change worth catching is a price line quietly taken off a shelf
// ticket.
func describeTemplate(in Template, fields string) map[string]any {
	return map[string]any{
		"name": in.Name, "kind": in.Kind,
		"width_mm": in.Width, "height_mm": in.Height,
		"is_default": in.IsDefault,
		"fields":     fields,
	}
}

// Sheet is what one print run needs: the layout, and a label per item.
type Sheet struct {
	Template Template `json:"template"`
	Labels   []Label  `json:"labels"`
}

// Query says which labels to produce.
//
// B3 asks for bulk generation by variant, by category and by brand — "a factory
// delivers 1,000 pieces → generate all 1,000 in one click" — and for a quantity
// per item, because a delivery of thirty of one shirt needs thirty tags.
type Query struct {
	TemplateID *uuid.UUID
	Kind       string
	VariantIDs []uuid.UUID
	CategoryID *uuid.UUID
	BrandID    *uuid.UUID
	Search     string
	// Copies repeats each label. Held here rather than left to the browser to
	// duplicate, so what is printed and what was asked for cannot diverge.
	Copies int
}

// Build assembles a print run.
func (s *Service) Build(ctx context.Context, scope Scope, q Query) (Sheet, error) {
	if q.Copies <= 0 {
		q.Copies = 1
	}
	if q.Copies > 100 {
		return Sheet{}, errs.New(errs.CodeInvalidInput,
			"A hundred copies of one label is the most this will print at once.")
	}

	var out Sheet
	err := s.pool.TxAsTenant(ctx, scope.TenantID, func(tx pgx.Tx) error {
		tpl, e := s.pickTemplate(ctx, tx, scope, q)
		if e != nil {
			return e
		}
		out.Template = tpl

		scheme, e := readScheme(ctx, tx, scope.CompanyID)
		if e != nil {
			return e
		}

		var currency, country string
		if e := tx.QueryRow(ctx,
			`SELECT base_currency, country FROM company WHERE id = $1`,
			scope.CompanyID).Scan(&currency, &country); e != nil {
			return e
		}

		// Resolved once for the run, on the transaction the caller already
		// holds. Asking the pool for a second connection while holding this one
		// is the deadlock that froze the tills; registry.Query.Tx exists so it
		// cannot happen again.
		rate, e := s.rules.VATRate(ctx, tx, country, time.Now().UTC(),
			scope.TenantID)
		if e != nil {
			return e
		}

		rows, e := tx.Query(ctx, labelSelect+`
			WHERE v.company_id = $1 AND v.is_active
			  AND ($2::uuid[] IS NULL OR v.id = ANY($2))
			  AND ($3::uuid IS NULL OR p.category_id = $3)
			  AND ($4::uuid IS NULL OR p.brand_id = $4)
			  AND ($5 = '' OR p.name ILIKE '%' || $5 || '%'
			                OR v.sku ILIKE '%' || $5 || '%')
			ORDER BY p.name, v.sku
			LIMIT 2000`,
			scope.CompanyID, nullIDs(q.VariantIDs), q.CategoryID, q.BrandID,
			strings.TrimSpace(q.Search))
		if e != nil {
			return e
		}
		defer rows.Close()

		out.Labels = []Label{}
		for rows.Next() {
			label, e := scanLabel(rows, scheme, currency, rate)
			if e != nil {
				return e
			}
			for i := 0; i < q.Copies; i++ {
				out.Labels = append(out.Labels, label)
			}
		}
		return rows.Err()
	})
	return out, db.Translate(err, "")
}

// pickTemplate resolves which layout to use.
func (s *Service) pickTemplate(
	ctx context.Context, tx pgx.Tx, scope Scope, q Query,
) (Template, error) {
	if q.TemplateID != nil {
		row := tx.QueryRow(ctx, templateSelect+`
			WHERE t.id = $1 AND t.company_id = $2`, *q.TemplateID, scope.CompanyID)
		tpl, err := scanTemplate(row)
		if err == pgx.ErrNoRows {
			return Template{}, errs.New(errs.CodeNotFound,
				"That label was not found.")
		}
		return tpl, err
	}

	kind := q.Kind
	if kind == "" {
		kind = KindThermal
	}
	row := tx.QueryRow(ctx, templateSelect+`
		WHERE t.company_id = $1 AND t.kind = $2
		ORDER BY t.is_default DESC, t.name LIMIT 1`, scope.CompanyID, kind)
	tpl, err := scanTemplate(row)
	if err == pgx.ErrNoRows {
		return Template{}, errs.Newf(errs.CodeNotFound,
			"This company has no %s label set up.", strings.ReplaceAll(kind, "_", " "))
	}
	return tpl, err
}

const templateSelect = `
	SELECT t.id, t.name, t.kind, t.width_mm, t.height_mm, t.columns, t.rows,
	       t.margin_mm, t.gap_mm, t.fields::text, t.is_default
	FROM label_template t`

// labelSelect reads everything a label prints, except the rate.
//
// The treatment travels with the row and the RATE is resolved once per print
// run through registry.VATRate — the single accessor every tax computation in
// this product goes through. Reading the registry table directly here would be
// a second definition of the rate, free to disagree with what the till charges,
// and a shelf ticket that disagrees with the till is the one error on this
// screen a customer notices and a regulator cares about.
const labelSelect = `
	SELECT v.id, v.sku, coalesce(v.barcode, ''), p.name,
	       coalesce(p.translations->>'ar', ''),
	       v.attributes::text,
	       coalesce(c.name, ''), coalesce(b.name, ''), coalesce(p.season, ''),
	       v.price_retail, p.tax_treatment
	FROM variant v
	JOIN product p ON p.id = v.product_id
	LEFT JOIN category c ON c.id = p.category_id
	LEFT JOIN brand b ON b.id = p.brand_id`

type scanner interface{ Scan(dest ...any) error }

func scanTemplate(row scanner) (Template, error) {
	var t Template
	var width, height, margin, gap decimal.Decimal
	var fields string
	if err := row.Scan(&t.ID, &t.Name, &t.Kind, &width, &height, &t.Columns,
		&t.Rows, &margin, &gap, &fields, &t.IsDefault); err != nil {
		return Template{}, err
	}
	if err := json.Unmarshal([]byte(fields), &t.Fields); err != nil {
		return Template{}, err
	}
	t.Width = width.String()
	t.Height = height.String()
	t.Margin = margin.String()
	t.Gap = gap.String()
	if t.Columns != nil && t.Rows != nil {
		t.PerSheet = *t.Columns * *t.Rows
	}
	return t, nil
}

func scanLabel(
	row scanner, scheme Scheme, currency string, standardRate decimal.Decimal,
) (Label, error) {
	var l Label
	var attributes, treatment string
	var price decimal.Decimal
	if err := row.Scan(&l.VariantID, &l.SKU, &l.Barcode, &l.Name, &l.NameAr,
		&attributes, &l.Category, &l.Brand, &l.Season, &price,
		&treatment); err != nil {
		return Label{}, err
	}

	// Only a standard-rated product carries the rate. A zero-rated or exempt
	// garment priced as though it were standard would put a tag on the rail
	// charging fifteen per cent tax the shop must not collect.
	rate := decimal.Zero
	if treatment == "standard" {
		rate = standardRate
	}

	var parsed map[string]string
	if err := json.Unmarshal([]byte(attributes), &parsed); err != nil {
		return Label{}, err
	}
	l.Attributes = describe(parsed)

	l.Symbology = scheme.Symbology
	l.Readable = buildReadable(scheme, l.Category, l.Brand, l.Season, parsed, l.SKU)
	if l.Barcode == "" {
		// Nothing has been generated for this variant yet. The SKU is printed
		// rather than an empty barcode, because a tag with a blank where the
		// bars should be is a tag somebody puts on a garment and then cannot
		// scan — and the SKU at least identifies it by eye.
		l.Barcode = l.SKU
	}

	// The shelf price, printed as it is held.
	//
	// `price_retail` is ALREADY VAT-inclusive in this product: the till prices
	// with PricesIncludeTax defaulting to true, so a 115.00 abaya rings up at
	// 115.00 and the 15.00 of tax inside it is derived by division. Adding the
	// rate here would print 132.25 on the tag of an item that sells for 115.00
	// — a wrong price physically attached to a garment, which is the one error
	// on this screen a customer notices at the counter.
	//
	// The rate travels alongside so a tag can say "VAT included", which B3
	// asks for and which is a caption rather than a calculation.
	l.Price = money(price.Round(2))
	l.Currency = currency
	l.TaxRate = rate.String()
	return l, nil
}

// describe turns a variant's attributes into the line under the name.
//
// Values only — "Black · XL" rather than "colour: Black · size: XL" — because
// the labels are what fits on 25 millimetres and a person looking at a black
// extra-large shirt does not need to be told which is which.
func describe(attributes map[string]string) string {
	// Colour and size first, in that order, because that is how clothing is
	// described out loud. Map order would put them differently on every label
	// printed, which is how a shelf of tags ends up unreadable.
	var parts []string
	for _, key := range []string{"colour", "color", "size"} {
		if v := strings.TrimSpace(attributes[key]); v != "" {
			parts = append(parts, v)
		}
	}
	for key, value := range attributes {
		switch key {
		case "colour", "color", "size":
			continue
		}
		if v := strings.TrimSpace(value); v != "" {
			parts = append(parts, v)
		}
	}
	return strings.Join(parts, " · ")
}

// PrintableFields are the things a label can carry, and the ONLY ones.
//
// Not a suggestion list: this is exactly the set the studio's renderer has a
// case for and the print run has data for. A template naming anything else
// produces a label with a silently missing line — the layout is right, the
// price is absent, and nobody discovers it until nine hundred tags are on a
// rail.
//
// So an unknown field is refused here rather than dropped, and the message
// names what there is. Adding a seventh means adding it to `Label`, to the
// print run's query, and to the renderer; a screen cannot introduce one by
// sending a word.
var PrintableFields = []string{
	"logo",          // the shop's mark
	"name",          // the product, as the catalogue holds it
	"name_ar",       // and its Arabic name, printed right to left
	"attributes",    // colour, size — the variant's own axes
	"price",         // VAT-inclusive, which is what a shelf ticket means
	"barcode",       // the bars, plus the human-readable string beneath
	"customer_name", // a loyalty card, which carries a person rather than a garment
}

func printable(field string) bool {
	for _, f := range PrintableFields {
		if f == field {
			return true
		}
	}
	return false
}

// checkFields refuses a field the printer cannot fill.
func checkFields(fields []Field) error {
	for i, f := range fields {
		name := strings.TrimSpace(f.Field)
		if printable(name) {
			continue
		}
		return errs.Validation("That label asks for something it cannot print.").
			WithField("fields",
				fmt.Sprintf("Line %d says %q. A label can carry: %s.",
					i+1, f.Field, strings.Join(PrintableFields, ", ")))
	}
	return nil
}

func knownKind(kind string) bool {
	switch kind {
	case KindHangTag, KindThermal, KindA4Sheet, KindLoyaltyCard:
		return true
	}
	return false
}

func amountOr(s string) decimal.Decimal {
	v, err := decimal.NewFromString(strings.TrimSpace(s))
	if err != nil {
		return decimal.Zero
	}
	return v
}

File: web-next/src/lib/labels/studio.ts
// Designing a label, and overriding the code on one product.
//
// # The field vocabulary is the backend's, not this screen's
//
// `labels.PrintableFields` names exactly what a print run can fill: a logo, the
// product name, its Arabic name, the variant's attributes, the VAT-inclusive
// price, the barcode, and — for a loyalty card — a customer name. The server
// refuses anything else, and this offers exactly the same list, so a template
// built here is one the printer can render.
//
// That refusal is the point. A label naming a field the run cannot fill prints
// a blank line and says nothing about it: the layout is right, the price is
// absent, and nobody discovers it until nine hundred tags are on a rail.
//
// # A sheet has a grid and a roll does not
//
// `label_template_sheet_has_a_grid` refuses columns and rows on anything but an
// A4 sheet, and refuses their absence on one. Stated here so the form asks for
// them where they belong rather than teaching the rule with a constraint
// violation.
//
// # Millimetres, because that is what the box says
//
// Every label roll and every printer driver is specified in them, and 50x25 is
// a number a shopkeeper reads off the packaging. Nothing here converts.

/** One line printed on a label. */
export interface LabelField {
  field: string;
  /** Point size, for text. */
  size?: number;
  /** Height in millimetres, for the barcode. */
  height?: number;
  bold?: boolean;
  rtl?: boolean;
}

/** A layout, as `GET /labels/templates` answers it. */
export interface LabelTemplate {
  id: string;
  name: string;
  kind: string;
  width_mm: string;
  height_mm: string;
  columns?: number;
  rows?: number;
  margin_mm: string;
  gap_mm: string;
  fields: LabelField[];
  is_default: boolean;
  /** columns x rows, which is how a shopkeeper buys the paper. */
  per_sheet?: number;
}

/** One label a print run produced, as `POST /labels/print` answers it. */
export interface PreparedLabel {
  variant_id: string;
  sku: string;
  barcode: string;
  symbology: string;
  /** The meaningful string, which for a digit symbology is not the barcode. */
  readable?: string;
  name: string;
  name_ar?: string;
  attributes?: string;
  category?: string;
  brand?: string;
  season?: string;
  /** VAT-inclusive: the shelf price is the price paid. */
  price: string;
  currency: string;
  tax_rate: string;
}

export interface PreparedSheet {
  template: LabelTemplate;
  labels: PreparedLabel[];
}

/** The kinds of label the schema admits. */
export const LABEL_KINDS = ['hang_tag', 'thermal', 'a4_sheet', 'loyalty_card'] as const;
export type LabelKind = (typeof LABEL_KINDS)[number];

/**
 * What a label can carry, in the order it reads down the sticker.
 *
 * The same seven `labels.PrintableFields` lists. `customer_name` is offered
 * only on a loyalty card, because that is the only kind whose subject is a
 * person rather than a garment — see `fieldsFor`.
 */
export const PRINTABLE_FIELDS = [
  'logo',
  'name',
  'name_ar',
  'attributes',
  'price',
  'barcode',
  'customer_name',
] as const;
export type PrintableField = (typeof PRINTABLE_FIELDS)[number];

/** Which of them make sense on this kind of label. */
export function fieldsFor(kind: string): PrintableField[] {
  return PRINTABLE_FIELDS.filter((f) =>
    kind === 'loyalty_card'
      ? f !== 'attributes' && f !== 'name_ar'
      : f !== 'customer_name',
  );
}

/** Fields that take a point size rather than a height in millimetres. */
export function isTextField(field: string): boolean {
  return field !== 'logo' && field !== 'barcode';
}

/** A layout being edited, before it becomes a request. */
export interface TemplateDraft {
  id?: string;
  name: string;
  kind: string;
  width_mm: string;
  height_mm: string;
  columns: string;
  rows: string;
  margin_mm: string;
  gap_mm: string;
  is_default: boolean;
  fields: LabelField[];
}

export type TemplateProblem =
  | 'no_name'
  | 'unknown_kind'
  | 'bad_width'
  | 'bad_height'
  | 'needs_grid'
  | 'no_fields'
  | 'unprintable_field'
  | null;

/** A positive decimal, as a millimetre measurement is. */
function positiveNumber(value: string): boolean {
  const text = value.trim();
  if (!/^\d+(\.\d+)?$/.test(text)) return false;
  return Number(text) > 0;
}

/**
 * What is still wrong with the layout, in the order somebody fixes it.
 *
 * Every one of these is a rule the server applies. Stating them here first
 * means the person learns the rule from the form rather than from a refusal;
 * the server stays the authority, so a rule that changes there becomes a
 * refusal here rather than a screen that quietly accepts something it should
 * not.
 */
export function templateProblem(draft: TemplateDraft): TemplateProblem {
  if (draft.name.trim() === '') return 'no_name';
  if (!(LABEL_KINDS as readonly string[]).includes(draft.kind)) return 'unknown_kind';
  if (!positiveNumber(draft.width_mm)) return 'bad_width';
  if (!positiveNumber(draft.height_mm)) return 'bad_height';
  if (draft.kind === 'a4_sheet') {
    const across = Number(draft.columns);
    const down = Number(draft.rows);
    if (!Number.isInteger(across) || across <= 0) return 'needs_grid';
    if (!Number.isInteger(down) || down <= 0) return 'needs_grid';
  }
  if (draft.fields.length === 0) return 'no_fields';
  if (
    draft.fields.some(
      (f) => !(PRINTABLE_FIELDS as readonly string[]).includes(f.field),
    )
  ) {
    return 'unprintable_field';
  }
  return null;
}

/**
 * The body the save route takes.
 *
 * A grid is sent only for an A4 sheet: the service nulls it for every other
 * kind, and sending one anyway would be a request that says something the
 * database refuses.
 */
export function templateBody(draft: TemplateDraft): Record<string, unknown> {
  const body: Record<string, unknown> = {
    name: draft.name.trim(),
    kind: draft.kind,
    width_mm: draft.width_mm.trim(),
    height_mm: draft.height_mm.trim(),
    margin_mm: draft.margin_mm.trim() || '0',
    gap_mm: draft.gap_mm.trim() || '0',
    is_default: draft.is_default,
    fields: draft.fields,
  };
  if (draft.kind === 'a4_sheet') {
    body.columns = Number(draft.columns);
    body.rows = Number(draft.rows);
  }
  return body;
}

/** An existing layout, as the editor holds it. */
export function draftOf(template: LabelTemplate): TemplateDraft {
  return {
    id: template.id,
    name: template.name,
    kind: template.kind,
    width_mm: template.width_mm,
    height_mm: template.height_mm,
    columns: template.columns ? String(template.columns) : '',
    rows: template.rows ? String(template.rows) : '',
    margin_mm: template.margin_mm,
    gap_mm: template.gap_mm,
    is_default: template.is_default,
    // Copied rather than shared: the editor mutates its own list, and a
    // reference into the cached response would change the table behind it.
    fields: template.fields.map((f) => ({ ...f })),
  };
}

/**
 * A blank layout of the given kind, with sizes the hardware actually comes in.
 *
 * Not a guess about what the shop wants — a guess about what the printer is.
 * 50x25 is what a thermal roll comes in and a sheet of 24 is what the box of
 * A4 labels says on it, so somebody starting a new layout does not have to
 * measure a sticker before they can save one.
 */
export function blankDraft(kind: LabelKind): TemplateDraft {
  const shape: Record<LabelKind, Partial<TemplateDraft>> = {
    thermal: { width_mm: '50', height_mm: '25', margin_mm: '1', gap_mm: '0' },
    hang_tag: { width_mm: '50', height_mm: '80', margin_mm: '3', gap_mm: '0' },
    a4_sheet: {
      width_mm: '63.5',
      height_mm: '33.9',
      columns: '3',
      rows: '8',
      margin_mm: '8',
      gap_mm: '2.5',
    },
    loyalty_card: {
      width_mm: '85.6',
      height_mm: '53.98',
      margin_mm: '4',
      gap_mm: '0',
    },
  };
  return {
    name: '',
    kind,
    width_mm: '',
    height_mm: '',
    columns: '',
    rows: '',
    margin_mm: '0',
    gap_mm: '0',
    is_default: false,
    fields:
      kind === 'loyalty_card'
        ? [
            { field: 'customer_name', size: 11, bold: true },
            { field: 'barcode', height: 12 },
          ]
        : [
            { field: 'name', size: 8 },
            { field: 'price', size: 10, bold: true },
            { field: 'barcode', height: 10 },
          ],
    ...shape[kind],
  };
}

/** How many labels one sheet holds. Null for a roll, which does not come in sheets. */
export function perSheet(t: {
  columns?: number | null;
  rows?: number | null;
}): number | null {
  if (!t.columns || !t.rows) return null;
  return t.columns * t.rows;
}

// ---------------------------------------------------------------------------
// Overriding one product's barcode
// ---------------------------------------------------------------------------

export type BarcodeProblem = 'empty' | 'unchanged' | 'taken_here' | null;

/**
 * What is wrong with a hand-assigned barcode, as far as this screen can tell.
 *
 * `taken_here` is a HINT, not the rule: it only sees the labels currently on
 * screen, and the shop has thousands. `variant_barcode_uq` is what actually
 * makes a code unique and the server's refusal is the authority — this exists
 * so an obvious clash is caught before somebody presses the button, not so the
 * check can be moved off the server.
 *
 * `unchanged` is not an error the server raises: it recognises the no-op and
 * records nothing. Saying so here stops somebody pressing a button that will
 * appear to do nothing.
 */
export function barcodeProblem(
  next: string,
  current: string,
  othersOnScreen: readonly { variant_id: string; barcode: string }[],
  variantID: string,
): BarcodeProblem {
  const code = next.trim();
  if (code === '') return 'empty';
  if (code === current.trim()) return 'unchanged';
  if (
    othersOnScreen.some((l) => l.variant_id !== variantID && l.barcode.trim() === code)
  ) {
    return 'taken_here';
  }
  return null;
}
